React/Views/RCTWKWebView.m

```diff
@@ -0,0 +1,291 @@
+  NSURL *url = request.URL;
+  NSString *scheme = url.scheme;
+
+  BOOL isJSNavigation = [scheme isEqualToString:RCTJSNavigationScheme];
+
+  // skip this for the JS Navigation handler
+  if (!isJSNavigation && _onShouldStartLoadWithRequest) {
+    NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+    [event addEntriesFromDictionary:@{
+      @"url" : (request.URL).absoluteString,
+      @"navigationType" : @(navigationAction.navigationType)
+    }];
+    if (![self.delegate webView:self
+            shouldStartLoadForRequest:event
+                         withCallback:_onShouldStartLoadWithRequest]) {
+      return decisionHandler(WKNavigationActionPolicyCancel);
+    }
+  }
+
+  if (_onLoadingStart) {
+    // We have this check to filter out iframe requests and whatnot
+    BOOL isTopFrame = [url isEqual:request.mainDocumentURL];
+    if (isTopFrame) {
+      NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+      [event addEntriesFromDictionary:@{
+        @"url" : url.absoluteString,
+        @"navigationType" : @(navigationAction.navigationType)
+      }];
+      _onLoadingStart(event);
+    }
+  }
+
+  if (isJSNavigation) {
+    // Handle messages posted
+    if ([request.URL.host isEqualToString:RCTJSPostMessageHost]) {
+      NSString *data = request.URL.query;
+      data = [data stringByReplacingOccurrencesOfString:@"+" withString:@" "];
+      data = [data stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+    
+      NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+      [event addEntriesFromDictionary: @{
+        @"data": data,
+      }];
+      _onMessage(event);
+    }
+  
+    return decisionHandler(WKNavigationActionPolicyCancel);
+  }
+
+  decisionHandler(WKNavigationActionPolicyAllow);
+}
+
+- (void)webView:(__unused WKWebView *)webView
+    didFailProvisionalNavigation:(__unused WKNavigation *)navigation
+                       withError:(NSError *)error
+{
+  if (_onLoadingError) {
+    if ([error.domain isEqualToString:NSURLErrorDomain] && error.code == NSURLErrorCancelled) {
+      // NSURLErrorCancelled is reported when a page has a redirect OR if you load
+      // a new URL in the WebView before the previous one came back. We can just
+      // ignore these since they aren't real errors.
+      // http://stackoverflow.com/questions/1024748/how-do-i-fix-nsurlerrordomain-error-999-in-iphone-3-0-os
+      return;
+    }
+
+    NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+    [event addEntriesFromDictionary:@{
+      @"domain" : error.domain,
+      @"code" : @(error.code),
+      @"description" : error.localizedDescription,
+    }];
+    _onLoadingError(event);
+  }
+}
+
+- (void)webView:(WKWebView *)webView
+    didFinishNavigation:(__unused WKNavigation *)navigation
+{
+  if (_injectedJavaScript != nil) {
+    [webView evaluateJavaScript:_injectedJavaScript completionHandler:^(id result, NSError *error) {
```

Comment:
I don't have a good understanding of `addUserScript` but it sounds to me that if the controller is instantiated on `initWithFrame` but scripts are added here, we'd just keep adding more scripts every time navigation finished.

I played around with adding it to `initWithFrame` but I don't think we have access to `_injectedJavascript` yet. 

Any ideas?